README.md
# fff

Dart package for convenient work with color. It will allow you to convert
colors between rgb, rgba and hex formats. Also it can parse color from
strings specified formats and from list of color components or map.

## Usage

### A simple usage Color:

```dart
import 'package:flutter_color_hex/color.dart';

main() {

    // create black color in rgb format
    const black = const Color();

    // or
    const black = const Color(0, 0, 0);

    // output it
    print(black); // rgb(0, 0, 0)
    print(black.toHexString()); // 000000
    print(black.toRgbString()); // rgb(0, 0, 0)
    print(black.toRgbaString()); // rgba(0, 0, 0, 1.0)

    // create red color in rgba format
    const red = const Color(255, 0, 0, 1.0);
}
```

Output color components as List object

```dart
import 'package:flutter_color_hex/color.dart';

main() {

    // create red color
    var red = const Color(255);

    // output simple list
    print(red.toList()); // [255, 0, 0]

    // output by template
    print(
      red.toList(template: [
        Component.ALPHA,
        Component.BLUE,
        Component.GREEN,
        Component.RED
      ])
    );
    // [1.0, 0, 0, 255]

    // value of the alpha channel always output as double and range is
    // set from 0 to 1 rgb components are always output as int. But
    // this can be changed by setting the parameter to true asDouble
    print(red.toList(asDouble: true)); // [255.0, 0.0, 0.0]

    // change of range of color components
    print(red.toList(range: 1)); // [1, 0, 0]

    // You can specify the number of times the output. For example it
    // may be useful to specify the color of each vertex polygon model
    print(red.toList(rows: 2, range: 1));
    // [1.0, 0.0, 0.0, 1.0, 0.0, 0.0]
}
```

Output color components as Map object

```dart
import 'package:flutter_color_hex/color.dart';

main() {

    var temp = {
      "A": Component.RED,
      "B": Component.GREEN,
      "C": Component.BLUE
    };

    // are all exactly the same as in the first case, with the only
    // exception that here there is no parameter rows and default
    // template {"red": RED, "green": GREEN, "blue": BLUE, "alpha":
    // ALPHA}
    print(Color(255, 100, 50).toMap(template: temp, range: 1));
    // {A: 1.0, B: 0.39215686274509803, C: 0.19607843137254902}

}
```

In addition and subtraction of colors, the left is the main color. That is, in the end will be the color of the same size as the left color. For color in rgb format value of alpha is 1.0
 
 ```dart
print(Color(100, 0, 0) + Color(50, 0, 0)); // rgb(150, 0, 0)
print(Color(100, 0, 0) - Color(50, 0, 0)); // rgb(50, 0, 0)

print(Color(100, 0, 0) + Color(50, 0, 0, .1)); // rgb(150, 0, 0)
print(Color(100, 0, 0) - Color(50, 0, 0, .1)); // rgb(50, 0, 0)

print(Color(100, 0, 0, .5) + Color(50, 0, 0)); // rgba(150, 0, 0, 1.0)
print(Color(100, 0, 0, .5) - Color(50, 0, 0)); // rgba(50, 0, 0, 0.0)

print(Color(100, 0, 0, .5) + Color(50, 0, 0, .3)); // rgba(150, 0, 0, 0.8)
print(Color(100, 0, 0, .5) - Color(50, 0, 0, .3)); // rgba(50, 0, 0, 0.2)
```

### Color presets

Presets library contains 275 constants of colors from [Material design](http://www.google.com/design/spec/style/color.html)

```dart
import 'package:flutter_color_hex/presets.dart';

main() {
    print(red + green); // rgb(255, 242, 134)
}
```

### Parse color

```dart
import 'package:flutter_color_hex/parser.dart';

main() {

    print(ColorParser("#FFF")); // rgb(255, 255, 255)
    print(ColorParser("#000000")); // rgb(0, 0, 0)
    print(ColorParser("#AAA") is Color); // true

    print(ColorParser("rgb(255, 255, 255)")); // rgb(255, 255, 255)
    print(ColorParser("rgba(255, 255, 255, 1.0)")); // rgba(255, 255, 255, 1.0)

    print(ColorParser([10, 20, 30])); // rgb(10, 20, 30)
    print(ColorParser([10, 20, 30, .4])); // rgba(10, 20, 30, 0.4)

    print(ColorParser({"r": 10, "g": 20, "b": 30, "a": .4})); // rgba(10, 20, 30, 0.4)
    print(ColorParser({"x": 10, "y": 20, "z": 30, "a": .4})); // rgba(10, 20, 30, 0.4)

}
```

### List convention

```dart
import 'package:flutter_color_hex/parser.dart';

main() {

    // wrong list of color components
    // but need rgb(10, 20, 30)
    var brg = [30, 10, 20];

    // change list convention
    listConvention = [
      Component.BLUE,
      Component.RED,
      Component.GREEN,
      Component.ALPHA
    ];

    print(ColorParser(brg)); // rgb(10, 20, 30)

}
```

### Map conventions

```dart
import 'package:flutter_color_hex/parser.dart';

main() {

    // by default set [rgba] and [xyza] conventions

    // add new          red   green  blue  alpha
    mapConventions.add(["A",  "B",   "C",  "D"]);

    print(ColorParser({"A": 10, "B": 20, "C": 30, "D": .9})); // rgba(10, 20, 30, 0.9)

    // Attention! Of all the conventions will be applied first best match.
    // Example:

    // remove all conventions
    mapConventions.clear();

    // add new
    mapConventions.add(["a", "b", "c", "A"]);
    mapConventions.add(["x", "y", "z", "A"]);
    mapConventions.add(["1", "2", "3", "A"]);

    // Notice: Need specify all four components (rgba).
    // Convention with the number of components other than the four will be ignored.

    // apply first convention (match "a", "b" and "c")
    print(ColorParser({"a": 10, "b": 20, "c": 30})); // rgb(10, 20, 30)

    // apply second convention (match "x", "y" and "z")
    print(ColorParser({"x": 10, "y": 20, "z": 30})); // rgb(10, 20, 30)

    // apply first convention (math "a" and "b")
    // "c" component is not found
    print(ColorParser({"a": 10, "b": 20, "z": 30})); // rgb(10, 20, 0)

    // apply second convention (math "y" and "z")
    // "x" component is not found
    print(ColorParser({"a": 10, "y": 20, "z": 30})); // rgb(0, 20, 30)

    print(ColorParser({"1": 10, "x": 20, "a": 30})); // rgb(30, 0, 0)

    // if the number of matches in several conventions is identical, then will be applied first

}
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/vdakalov/fff/issues


CHANGELOG.md
# Changelog

## 2.0.0

- Add null-safety

## 1.0.1

- Support for dart 2.0.0 SDK

## 1.0.0

- Change self-enum color components with Dart builtin enums
- Remove unused enum items (RGB, RGBA, HEX)
- Move color components from `package:flutter_color_hex/parser.dart` to `package:flutter_color_hex/color.dart`
- Restored tests to work with the new Dart library
- Added the ability to specify a pattern when translating a color into a string

## 0.0.3

- Added method toList for the Color object
- Update docs for constants in presets library
- Added method toMap for the Color object

## 0.0.2+1

- Remove excess annotations types
- Fix changelog version

## 0.0.2

- Simple Color constructor was changed to constant constructor
- Drew parse method in a separate factory ColorParse
- Add 275 preset [Material design](http://www.google.com/design/spec/style/color.html)
- Add more tests

## 0.0.1

- Implements Color class


test/color_test.dart
// Copyright (c) 2015, Viktor Dakalov. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

library fff.test.color;

import 'package:flutter_color_hex/color.dart';
import 'package:test/test.dart';

main() {
  test("Constructor without arguments", () {
    expect(ColorHex().toString(), "rgb(0, 0, 0)");
  });

  test("Constructor with arguments", () {
    expect(ColorHex(255).toString(), "rgb(255, 0, 0)");
    expect(ColorHex(0, 255).toString(), "rgb(0, 255, 0)");
    expect(ColorHex(0, 0, 255).toString(), "rgb(0, 0, 255)");
    expect(ColorHex(0, 0, 0, .3).toString(), "rgba(0, 0, 0, 0.3)");
    expect(ColorHex(1, 2, 3, 0.4).toString(), "rgba(1, 2, 3, 0.4)");
  });

  test("Output format", () {
    expect(ColorHex(255).toString(), "rgb(255, 0, 0)");
    expect(ColorHex(255).toRgbString(), "rgb(255, 0, 0)");

    expect(ColorHex(255, 0, 0, 1.0).toString(), "rgba(255, 0, 0, 1.0)");
    expect(ColorHex(255, 0, 0, 1.0).toRgbaString(), "rgba(255, 0, 0, 1.0)");

    expect(ColorHex(255, 0, 0).toHexString(), "ff0000");

    expect(ColorHex(255, 255, 255).toString('red: %r, green: %g, blue: %B'),
        'red: 255, green: 255, blue: 255');

    expect(
        ColorHex(255, 255, 255, .5)
            .toString('red: %r, green: %g, blue: %B, alpha: %A'),
        'red: 255, green: 255, blue: 255, alpha: 0.5');

    expect(ColorHex(255, 255, 255, .5).toHexString('#%R%g%b'), '#FFffff');

    // === toList ===

    // simple
    expect(ColorHex(0, 0, 0).toList(), equals([0, 0, 0]));
    expect(ColorHex(255, 255, 255).toList(), equals([255, 255, 255]));
    expect(ColorHex(0, 0, 0, 0.0).toList(), equals([0, 0, 0, 0.0]));
    expect(ColorHex(255, 255, 255, 1.0).toList(), equals([255, 255, 255, 1.0]));

    // as double
    expect(ColorHex(10, 20, 30).toList(asDouble: true),
        equals([10.0, 20.0, 30.0]));
    expect(ColorHex(10, 20, 30, 0.4).toList(asDouble: true),
        equals([10.0, 20.0, 30.0, 0.4]));

    // template
    expect(
        ColorHex(10, 20, 30, 0.4).toList(template: [
          "A",
          Component.RED,
          "B",
          Component.GREEN,
          "C",
          Component.BLUE,
          "D",
          Component.ALPHA
        ]),
        equals(["A", 10.0, "B", 20.0, "C", 30.0, "D", 0.4]));

    expect(ColorHex(10, 20, 30).toList(template: ["A", "B", "C"]),
        equals(["A", "B", "C"]));

    expect(
        ColorHex(10, 20, 30).toList(
            template: [Component.ALPHA, Component.ALPHA, Component.ALPHA]),
        equals([1.0, 1.0, 1.0]));

    expect(
        ColorHex(10, 20, 30)
            .toList(template: [Component.BLUE, Component.BLUE, Component.BLUE]),
        equals([30, 30, 30]));

    // range
    expect(
        ColorHex(50, 255, 100, 0.4)
            .toList(asDouble: false, range: 1, template: [
          Component.RED,
          Component.GREEN,
          Component.BLUE,
          Component.ALPHA,
          "A",
          Component.RED,
          "B",
          Component.GREEN,
          "C",
          Component.BLUE,
          "D",
          Component.ALPHA
        ]),
        equals([
          0.19607843137254902,
          1.0,
          0.39215686274509803,
          0.4,
          "A",
          0.19607843137254902,
          "B",
          1.0,
          "C",
          0.39215686274509803,
          "D",
          0.4
        ]));

    expect(ColorHex(10, 20, 30).toList(rows: 3),
        equals([10, 20, 30, 10, 20, 30, 10, 20, 30]));

    // === toMap ===

    expect(ColorHex(10, 20, 30).toMap(),
        equals({"red": 10, "green": 20, "blue": 30}));
    expect(ColorHex(10, 20, 30, 0.0).toMap(),
        equals({"red": 10, "green": 20, "blue": 30, "alpha": 0.0}));

    // as double
    expect(ColorHex(10, 20, 30).toMap(asDouble: true),
        equals({"red": 10.0, "green": 20.0, "blue": 30.0}));
    expect(ColorHex(10, 20, 30, 0.6).toMap(asDouble: true),
        equals({"red": 10.0, "green": 20.0, "blue": 30.0, "alpha": 0.6}));

    // template
    expect(
        ColorHex(10, 20, 30).toMap(template: {
          "a": Component.RED,
          "b": Component.GREEN,
          "c": Component.BLUE
        }),
        equals({"a": 10, "b": 20, "c": 30}));
    expect(ColorHex(10, 20, 30).toMap(template: {}), equals({}));
    expect(
        ColorHex(10, 20, 30).toMap(template: {
          "a": Component.RED,
          "A": Component.RED,
          "b": Component.GREEN,
          "B": Component.GREEN,
          "c": Component.BLUE,
          "C": Component.BLUE,
          "D": Component.ALPHA
        }),
        equals(
            {"a": 10, "A": 10, "b": 20, "B": 20, "c": 30, "C": 30, "D": 1.0}));

    // range
    expect(
        ColorHex(50, 255, 100, 0.4).toMap(template: {
          "a": Component.RED,
          "b": Component.GREEN,
          "c": Component.BLUE
        }, range: 1),
        equals({"a": 0.19607843137254902, "b": 1.0, "c": 0.39215686274509803}));
  });

  test("Arithmetics", () {
    expect(ColorHex(255) + ColorHex(0, 255), ColorHex(255, 255));
    expect(ColorHex(255) - ColorHex(100), ColorHex(155, 0, 0));
  });
}


test/utils_test.dart
// Copyright (c) 2015, Viktor Dakalov. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

library fff.test.utils;

import 'package:flutter_color_hex/utils.dart';
import 'package:test/test.dart';

main() {
  test("min", () {
    expect(min(10, 20), 10);
    expect(min(-2, -1), -2);
  });

  test("max", () {
    expect(max(100, 0), 100);
    expect(max(-20, -19.9), -19.9);
  });

  test("hexin", () {
    expect(hexin("f"), "ff");
    expect(hexin("ff"), "ff");
  });

  test("hexout", () {
    expect(hexout("f"), "0f");
    expect(hexout("ff"), "ff");
  });

  test("tohex", () {
    expect(tohex(255), "ff");
    expect(tohex(0), "00");
    expect(tohex(240), "f0");
  });

  test("fromhex", () {
    expect(fromhex("ff"), 255);
    expect(fromhex("0"), 0);
  });
}


test/color_parser_test.dart
// Copyright (c) 2015, Viktor Dakalov. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

library fff.test.color_parser;

import 'package:flutter_color_hex/color.dart';
import 'package:flutter_color_hex/parser.dart';
import 'package:test/test.dart';

main() {
  test("ColorParser from HEX", () {
    expect(ColorParser("#fff").toString(), "rgb(255, 255, 255)");
    expect(ColorParser("#FFF").toString(), "rgb(255, 255, 255)");
    expect(ColorParser("#fafafa").toString(), "rgb(250, 250, 250)");
    expect(ColorParser("#FAFAFA").toString(), "rgb(250, 250, 250)");
  });

  test("ColorParser fromList", () {
    expect(ColorParser([1]).toString(), "rgb(1, 0, 0)");
    expect(ColorParser([1, 2]).toString(), "rgb(1, 2, 0)");
    expect(ColorParser([1, 2, 3]).toString(), "rgb(1, 2, 3)");
    expect(ColorParser([1, 2, 3, .4]).toString(), "rgba(1, 2, 3, 0.4)");

    expect(ColorParser([10, null, null, null]).toString(), "rgb(10, 0, 0)");
    expect(ColorParser([null, 10, null, null]).toString(), "rgb(0, 10, 0)");
    expect(ColorParser([null, null, 10, null]).toString(), "rgb(0, 0, 10)");
    expect(
        ColorParser([null, null, null, .1]).toString(), "rgba(0, 0, 0, 0.1)");
  });

  test("ColorParser fromMap", () {
    expect(ColorParser({"r": 1}).toString(), "rgb(1, 0, 0)");
    expect(ColorParser({"g": 2}).toString(), "rgb(0, 2, 0)");
    expect(ColorParser({"b": 3}).toString(), "rgb(0, 0, 3)");
    expect(ColorParser({"a": .4}).toString(), "rgba(0, 0, 0, 0.4)");
    expect(ColorParser({"r": 1, "g": 2, "b": 3, "a": .4}).toString(),
        "rgba(1, 2, 3, 0.4)");
  });

  test("List convention", () {
    var def = listConvention;

    listConvention = [
      Component.ALPHA,
      Component.BLUE,
      Component.GREEN,
      Component.RED
    ];
    expect(ColorParser([.1, 3, 2, 1]).toString(), "rgba(1, 2, 3, 0.1)");

    listConvention = [Component.BLUE, Component.ALPHA];
    expect(ColorParser([3, .2]).toString(), "rgba(0, 0, 3, 0.2)");

    listConvention = def;
  });

  test("Map convensions", () {
    var def = mapConventions;

    mapConventions.clear();
    expect(() => ColorParser({"R": 1, "G": 2, "B": 3, "A": .1}),
        throwsA(isException));

    mapConventions.add(["X", "Y", "Z"]);
    mapConventions.add(["A", "B", "C", "D"]);
    mapConventions.add(["X", "Y", "Z", "A"]);
    mapConventions.add(["R", "G", "B", "A"]);

    expect(ColorParser({"X": 1, "Y": 2, "Z": 3, "T": .4}).toString(),
        "rgb(1, 2, 3)");
    expect(ColorParser({"A": 2, "B": 4, "C": 6, "D": .8}).toString(),
        "rgba(2, 4, 6, 0.8)");
    expect(ColorParser({"X": 1, "Y": 2, "Z": 3, "A": .4}).toString(),
        "rgba(1, 2, 3, 0.4)");
    expect(ColorParser({"R": 1, "G": 2, "B": 3, "A": .4}).toString(),
        "rgba(1, 2, 3, 0.4)");

    mapConventions = def;
  });

  test("Parse color", () {
    var rgb = "rgb(240, 240, 240)",
        rgba = "rgba(240, 240, 240, 1.0)",
        hex = "#f0f0f0",
        hex2 = "f0f0f0";

    expect(ColorParser(rgb).toRgbString(), rgb);
    expect(ColorParser(rgb).toRgbaString(), rgba);
    expect(ColorParser(rgb).toHexString(), hex2);

    expect(ColorParser(rgba).toRgbString(), rgb);
    expect(ColorParser(rgba).toRgbaString(), rgba);
    expect(ColorParser(rgba).toHexString(), hex2);

    expect(ColorParser(hex).toRgbString(), rgb);
    expect(ColorParser(hex).toRgbaString(), rgba);
    expect(ColorParser(hex).toHexString(), hex2);
  });

  test("Default values for ColorParser", () {
    DEF_RED = 50;
    DEF_GREEN = 50;
    DEF_BLUE = 50;
    DEF_ALPHA = .5;

    expect(() => ColorParser(), throwsA(isException));
    expect(ColorParser(100).toString(), "rgba(100, 50, 50, 0.5)");
    expect(ColorParser(null, 100).toString(), "rgba(50, 100, 50, 0.5)");
    expect(ColorParser(null, null, 100).toString(), "rgba(50, 50, 100, 0.5)");
    expect(
        ColorParser(null, null, null, .6).toString(), "rgba(50, 50, 50, 0.6)");

    DEF_RED = 0;
    DEF_GREEN = 0;
    DEF_BLUE = 0;
    DEF_ALPHA = null;
  });

  test("Arithmetics", () {
    expect(ColorHex(255) == ColorParser("#f00"), isTrue);
    expect(
        ColorHex(255, 255, 255) == ColorParser("rgb(255, 255, 255)"), isTrue);

    expect(ColorHex(255, 255, 255) == ColorParser("rgba(255, 255, 255, 1.0)"),
        isFalse);
    expect(ColorHex(255, 255, 255, 1.0) == ColorParser("#FFF"), isFalse);
  });

  test("Restrictions", () {
    expect(ColorParser(512), ColorHex(255));
    expect(ColorParser(-1024), ColorHex(0));
  });
}


lib/color.dart
// Copyright (c) 2015, Viktor Dakalov. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

library color;

import 'utils.dart';

enum Component { RED, GREEN, BLUE, ALPHA }

class ColorHex {
  /// Return red component (0-255)
  final int red;

  /// Return green component (0-255)
  final int green;

  /// Return blue component (0-255)
  final int blue;

  /// Return opacity level (null or 0-1)
  final double? alpha;

  /// Constructor create [Color] from rgba components.
  /// Default value for [red], [green] and [blue] component is 0, for [alpha]- null
  /// For color with specified [alpha] component output format will be rgba.
  const ColorHex([this.red = 0, this.green = 0, this.blue = 0, this.alpha]);

  bool operator ==(Object other) {
    if (other is ColorHex) {
      return red == other.red &&
          green == other.green &&
          blue == other.blue &&
          alpha == other.alpha;
    }
    return false;
  }

  ColorHex operator +(ColorHex other) {
    return ColorHex(
        min(red + other.red, 255),
        min(green + other.green, 255),
        min(blue + other.blue, 255),
        alpha == null
            ? alpha
            : min(alpha! + (other.alpha == null ? 1.0 : other.alpha!), 1.0));
  }

  ColorHex operator -(ColorHex other) {
    return ColorHex(
        max(red - other.red, 0),
        max(green - other.green, 0),
        max(blue - other.blue, 0),
        alpha == null
            ? alpha
            : max(alpha! - (other.alpha == null ? 1.0 : other.alpha!), 0.0));
  }

  /// Output color as string in rgba(r, g, b, a) or rgb(r, g, b) format
  String toString([template = null]) => alpha == null
      ? (template == null ? toRgbString() : toRgbString(template))
      : (template == null ? toRgbaString() : toRgbaString(template));

  /// Output color in rgba format
  String toRgbaString([String template = 'rgba(%r, %g, %b, %a)']) => template
      .replaceAll(RegExp('%[rR]'), red.toString())
      .replaceAll(RegExp('%[gG]'), green.toString())
      .replaceAll(RegExp('%[bB]'), blue.toString())
      .replaceAll(RegExp('%[aA]'), (alpha == null ? 1.0 : alpha).toString());

  /// Output color in rgb format
  String toRgbString([String template = 'rgb(%r, %g, %b)']) => template
      .replaceAll(RegExp('%[rR]'), red.toString())
      .replaceAll(RegExp('%[gG]'), green.toString())
      .replaceAll(RegExp('%[bB]'), blue.toString());

  /// Output color in hex format
  String toHexString([String template = '%r%g%b']) => template
      .replaceAll(RegExp('%r'), tohex(red))
      .replaceAll(RegExp('%R'), tohex(red).toUpperCase())
      .replaceAll(RegExp('%g'), tohex(green))
      .replaceAll(RegExp('%G'), tohex(green).toUpperCase())
      .replaceAll(RegExp('%b'), tohex(blue))
      .replaceAll(RegExp('%B'), tohex(blue).toUpperCase());

  /// Output color as [List] object
  /// By default, the list will be output in rgb or rgba format, depending on
  /// the format of this object. You can also specify an arbitrary number and
  /// order of the components are specify in the [template]. Default values rgb
  /// component output as int. You can change this bypassing a true argument
  /// [asDouble]. All rgb components of the default set in the range of 0-255,
  /// but you can change this by specifying a range argument [range] from 0-range
  /// (eg from 0 to 1). It is also possible to deduce the components several
  /// times, specify the argument [rows]. This may be necessary when you need to
  /// set the same color for each vertex in the polygon 3D model.
  List toList({List? template, bool asDouble: false, int? range, int rows: 1}) {
    var values = [red, green, blue, alpha],
        defEnd = 255,
        end = range is int ? range : defEnd,
        comps = template == null && alpha == null
            ? [Component.RED, Component.GREEN, Component.BLUE]
            : [Component.RED, Component.GREEN, Component.BLUE, Component.ALPHA],
        target = template == null ? List.from(comps) : List.from(template);

    for (var index = 0; index < comps.length;) {
      var pos = target.indexOf(comps[index]),
          isAlpha = comps[index] == Component.ALPHA,
          value = values[index];
      if (pos is int && pos >= 0) {
        if (isAlpha) {
          value = value == null ? 1.0 : value;
        }
        if (value is int && asDouble) {
          value = value.toDouble();
        }
        if (value is int && end != defEnd && !isAlpha) {
          value = range! / defEnd * value;
        }
        target[pos] = value;
      } else {
        index++;
      }
    }

    if (rows > 1) {
      var _target = List.from(target);
      while (rows-- > 1) {
        target.addAll(_target);
      }
    }

    return target;
  }

  /// Output color as [Map] object
  /// By default, the map will be output with the keys "red", "green" and
  /// "blue". "alpha" will be displayed depending on the format of the object.
  /// You can also specify an arbitrary number of components of said [template].
  /// Default values rgb component output as int. You can change this by
  /// passing a true argument [asDouble]. All rgb components of the default set
  /// in the range of 0-255, but you can change this by specifying a [range]
  /// argument needs range from 0-range (eg from 0 to 1).
  Map toMap({Map? template, bool asDouble: false, int? range}) {
    var defKeys = ["red", "green", "blue", "alpha"],
        comps = template == null && alpha == null
            ? [Component.RED, Component.GREEN, Component.BLUE]
            : [Component.RED, Component.GREEN, Component.BLUE, Component.ALPHA],
        target = template is Map
            ? template
            : Map.fromIterables(defKeys.sublist(0, comps.length), comps);

    return Map.fromIterables(
        target.keys,
        toList(
            template: target.values.toList(),
            asDouble: asDouble,
            range: range));
  }
}


lib/parser.dart
// Copyright (c) 2015, Viktor Dakalov. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

library color_parser;

import 'color.dart';
import 'utils.dart';

/// Default value of red component
num DEF_RED = 0;

/// Default value of green component
num DEF_GREEN = 0;

/// Default value of blue component
num DEF_BLUE = 0;

/// Default value of alpha component
num? DEF_ALPHA = null;

/// Agreement on the map color format
List<List<String>> mapConventions = [
  ['r', 'g', 'b', 'a'],
  ['x', 'y', 'z', 'a']
];

/// Agreement on the list color format
List listConvention = [
  Component.RED,
  Component.GREEN,
  Component.BLUE,
  Component.ALPHA
];

_parseComponents(String input, String prefix, [String postfix = ")"]) {
  if (input.startsWith(prefix) && input.contains(postfix)) {
    return List.from(input
        .substring(prefix.length, input.indexOf(postfix))
        .split(",")
        .map((c) => num.parse(c))
        .toList());
  }
  return [];
}

_parseRgb(String input) {
  var comps = _parseComponents(input, "rgb(");
  return [
    comps.length > 0 ? comps[0] : DEF_RED,
    comps.length > 1 ? comps[1] : DEF_GREEN,
    comps.length > 2 ? comps[2] : DEF_BLUE
  ];
}

_parseRgba(String input) {
  var comps = _parseComponents(input, "rgba(");
  return [
    comps.length > 0 ? comps[0] : DEF_RED,
    comps.length > 1 ? comps[1] : DEF_GREEN,
    comps.length > 2 ? comps[2] : DEF_BLUE,
    comps.length > 3 ? comps[3] : 1.0
  ];
}

_parseHex(String input) {
  if (input.length == 3) {
    return [input.substring(0, 1), input.substring(1, 2), input.substring(2, 3)]
        .map(fromhex)
        .toList();
  }

  if (input.length == 6) {
    return [input.substring(0, 2), input.substring(2, 4), input.substring(4, 6)]
        .map(fromhex)
        .toList();
  }

  return [];
}

_parseList(List<num?> input) {
  int ri = listConvention.indexOf(Component.RED),
      gi = listConvention.indexOf(Component.GREEN),
      bi = listConvention.indexOf(Component.BLUE),
      ai = listConvention.indexOf(Component.ALPHA);

  return [
    ri >= 0 && input.length > ri ? input[ri] : null,
    gi >= 0 && input.length > gi ? input[gi] : null,
    bi >= 0 && input.length > bi ? input[bi] : null,
    ai >= 0 && input.length > ai ? input[ai] : null
  ];
}

_parseMap(Map input) {
  var conventions = List.from(mapConventions).reversed.toList(),
      convention,
      result = [];

  if (conventions.length == 0) {
    throw Exception(
        "To create an object of Color from data a map type, you need specify one or more map conventions");
  }

  for (var index = 0; index < conventions.length; index++) {
    var convention = conventions[index], c = convention.length;

    if (convention.length != 4) {
      continue;
    }

    for (var j = 0; j < convention.length; j++) {
      if (input.containsKey(convention[j])) {
        c--;
      }
    }

    if (c >= result.length) {
      result.length = c + 1;
    }

    result.insert(c, convention);
  }

  if (result.length == 0) {
    throw Exception(
        "The properties of the map does not match any of the conventions");
  }

  convention = result.firstWhere((c) => c is List);

  return [
    input[convention[0]],
    input[convention[1]],
    input[convention[2]],
    input[convention[3]]
  ];
}

_parseArgs(List args) {
  return [
    args.length > 0 && args[0] is num
        ? max(min(args[0].toInt(), 255), 0)
        : DEF_RED,
    args.length > 1 && args[1] is num
        ? max(min(args[1].toInt(), 255), 0)
        : DEF_GREEN,
    args.length > 2 && args[2] is num
        ? max(min(args[2].toInt(), 255), 0)
        : DEF_BLUE,
    args.length > 3 && args[3] is num
        ? max(min(args[3].toDouble(), 1.0), 0.0)
        : DEF_ALPHA
  ];
}

/// Factory for Color class
/// Default values for rgba components specify in [DEF_RED], [DEF_GREEN], [DEF_BLUE] and [DEF_ALPHA]
///
/// allow follow color format:
///
/// - rgba(255, 255, 255, 0.2)
/// - rgb(255, 255, 255)
/// - \#FfF
/// - \#FaFaFa
///
/// return [Color] object
///
///
///  expect(ColorParser([10, null, null, null]).toString(), "rgb(10, 0, 0)");
ColorHex ColorParser([dynamic? red, num? green, num? blue, num? alpha]) {
  var args;

  if (red is String) {
    if (red.contains(")")) {
      if (red.startsWith("rgba(")) {
        args = _parseRgba(red);
      } else if (red.startsWith("rgb(")) {
        args = _parseRgb(red);
      }
    } else if (red.startsWith("#")) {
      args = _parseHex(red.substring(1));
    }
  } else if (red is List<num?>) {
    args = _parseList(red);
  } else if (red is Map<dynamic, num>) {
    args = _parseMap(red);
  } else if (red is num || green is num || blue is num || alpha is num) {
    args = [red, green, blue, alpha];
  }

  if (args == null) {
    throw Exception(
        "Invalid color format. Allowed formats: #FFF, #FFFFFF, rgba(255, 255, 255, 1.0) and rgb(255, 255, 255)");
  }

  args = _parseArgs(args);

  return ColorHex(args[0], args[1], args[2], args[3]);
}
